File: app/src/main/java/com/example/sunkai/heritage/ConnectWebService/HandleFind.java
package com.example.sunkai.heritage.ConnectWebService;

import android.support.annotation.Nullable;

import com.example.sunkai.heritage.Data.FindActivityAllData;
import com.example.sunkai.heritage.Data.FindActivityData;
import com.example.sunkai.heritage.Data.commentReplyData;
import com.example.sunkai.heritage.Data.userCommentData;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.kobjects.base64.Base64;
import org.ksoap2.serialization.SoapObject;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by sunkai on 2017-4-24.
 * 此类封装了发现页相关服务器请求的方法
 */

public class HandleFind extends BaseSetting {
    @Nullable
    private static List<userCommentData> Json_To_UserCommentData(String json) {
        try {
            if (null == json) {
                return null;
            }
            JSONObject MainActivity = new JSONObject(json);
            JSONArray activities = MainActivity.getJSONArray("user_comment_information");
            List<userCommentData> folkInformations = new ArrayList<>();
            for (int i = 0; i < activities.length(); i++) {
                userCommentData data = new userCommentData();
                JSONObject activity = (JSONObject) activities.get(i);
                data.setId((int) activity.get("id"));
                data.setUser_id((int) activity.get("user_id"));
                data.setCommentTime((String) activity.get("comment_time"));
                data.setCommentTitle((String) activity.get("comment_title"));
                data.setCommentContent((String) activity.get("comment_content"));
                data.setCommentLikeNum((String) activity.get("comment_num"));
                data.setCommentReplyNum((String) activity.get("reply_num"));
                data.setUserName((String) activity.get("user_name"));
                data.setUserFocusUser(success.equals((String)activity.get("isFollow")));
                data.setUserLike(success.equals((String)activity.get("isLike")));
                folkInformations.add(data);
            }
            return folkInformations;
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return null;
    }

    @Nullable
    private static List<commentReplyData> Json_To_UserCommentReplyData(String json) {
        try {
            JSONObject reply = new JSONObject(json);
            JSONArray replys = reply.getJSONArray("reply_information");
            List<commentReplyData> datas = new ArrayList<>();
            for (int i = 0; i < replys.length(); i++) {
                commentReplyData data = new commentReplyData();
                JSONObject oneReply = (JSONObject) replys.get(i);
                data.setReplyId((int) oneReply.get("reply_id"));
                data.setReplyTime((String) oneReply.get("reply_time"));
                data.setReplyContent((String) oneReply.get("reply_content"));
                data.setUserName((String) oneReply.get("user_name"));
                datas.add(data);
            }
            return datas;

        } catch (JSONException e) {
            e.printStackTrace();
        }
        return null;
    }

    @Nullable
    private static List<FindActivityData> Json_To_FindAcitivityID(List<FindActivityData> datas, String json) {
        if(error.equals(json)||json==null)
            return null;
        try {
            JSONObject js = new JSONObject(json);
            JSONArray replys = js.getJSONArray("id");
            for (int i = 0; i < replys.length(); i++) {
                FindActivityData data = new FindActivityData();
                data.setId( (int) replys.get(i));
                datas.add(data);
            }
            return datas;
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return null;
    }

    private static FindActivityData Json_To_FindActivityInformation(int id, String json) {
        try {
            FindActivityData getdata = new FindActivityData();
            getdata.setId(id);
            JSONObject js = new JSONObject(json);
            getdata.setTitle(js.getString("title"));
            getdata.setContent(js.getString("content"));
            String imgCode = js.getString("image");
            getdata.setImage(Base64.decode(imgCode));
            return getdata;
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return null;
    }

    private static List<Integer> Json_To_FindComment_ID(String json){
        try{
            JSONObject js=new JSONObject(json);
            JSONArray jsonArray=js.getJSONArray("members");
            List<Integer> ids=new ArrayList<>();
            for(int i=0;i<jsonArray.length();i++){
                ids.add((int)jsonArray.get(i));
            }
            return ids;
        }catch (JSONException e){
            e.printStackTrace();
            return null;
        }
    }

    private static FindActivityAllData Json_To_FindActivityAllInformation(String json){
        try{
            JSONObject js=new JSONObject(json);
            FindActivityAllData data=new FindActivityAllData();
            data.setId(js.getInt("id"));
            data.setUserName(js.getString("userName"));
            data.setUserID(js.getInt("userID"));
            data.setComent_time(js.getString("comment_time"));
            data.setComment_title(js.getString("comment_title"));
            data.setComment_content(js.getString("comment_content"));
            data.setReplyCount(js.getString("replyCount"));
            data.setUserLike(success.equals(js.getString("isUserLike")));
            data.setLikeNumber(js.getString("likeNumber"));
            data.setUserFlow(success.equals(js.getString("isUserFllow")));
            data.setImgCode(js.getString("image"));
            return data;
        }catch (JSONException e){
            e.printStackTrace();
            return null;
        }
    }

    public static List<FindActivityData> Get_Find_Activity_ID(List<FindActivityData> datas) {
        methodName = "Get_Find_Activity_ID";
        soapAction = namespace + "/" + methodName;
        SoapObject soapObject = new SoapObject(namespace, methodName);
        String result = Get_Post(soapObject);
        return Json_To_FindAcitivityID(datas, result);
    }

    public static FindActivityData Get_Find_Activity_Information(int id) {
        methodName = "Get_Find_Activity_Information";
        soapAction = namespace + "/" + methodName;
        SoapObject soapObject = new SoapObject(namespace, methodName);
        soapObject.addProperty("id", id);
        String result = Get_Post(soapObject);
        return Json_To_FindActivityInformation(id, result);
    }

    public static boolean Add_User_Comment_Information(int user_id, String comment_title, String comment_content, String comment_image) {
        methodName = "Add_User_Comment_Information";
        soapAction = namespace + "/" + methodName;
        SoapObject soapObject = new SoapObject(namespace, methodName);
        soapObject.addProperty("user_id", user_id);
        soapObject.addProperty("comment_title", comment_title);
        soapObject.addProperty("comment_content", comment_content);
        soapObject.addProperty("comment_image", comment_image);
        String result = Get_Post(soapObject);
        if (success.equals(result)) {
            return true;
        } else {
            return false;
        }
    }

    public static List<userCommentData> Get_User_Comment_Information(int userID) {
        methodName = "Get_User_Comment_Information";
        soapAction = namespace + "/" + methodName;
        SoapObject soapObject = new SoapObject(namespace, methodName);
        soapObject.addProperty("userID",userID);
        String result = Get_Post(soapObject);
        return Json_To_UserCommentData(result);
    }

    public static List<userCommentData> Get_User_Comment_Information_By_User(int userID) {
        methodName = "Get_User_Comment_Information_By_User";
        soapAction = namespace + "/" + methodName;
        SoapObject soapObject = new SoapObject(namespace, methodName);
        soapObject.addProperty("userID", userID);
        String result = Get_Post(soapObject);
        return Json_To_UserCommentData(result);
    }

    public static List<userCommentData> Get_User_Comment_Information_By_Own(int userID) {
        methodName = "Get_User_Comment_Information_By_Own";
        soapAction = namespace + "/" + methodName;
        SoapObject soapObject = new SoapObject(namespace, methodName);
        soapObject.addProperty("userID", userID);
        String result = Get_Post(soapObject);
        return Json_To_UserCommentData(result);
    }

    public static List<Integer> Get_User_Comment_ID(){
        methodName="Get_User_Comment_ID";
        soapAction = namespace + "/" + methodName;
        SoapObject soapObject = new SoapObject(namespace, methodName);
        String result=Get_Post(soapObject);
        if(error.equals(result))
            return null;
        return Json_To_FindComment_ID(result);
    }

    public static List<Integer> Get_User_Comment_ID_By_User(int userID){
        methodName="Get_User_Comment_ID_By_User";
        soapAction = namespace + "/" + methodName;
        SoapObject soapObject = new SoapObject(namespace, methodName);
        soapObject.addProperty("userID",userID);
        String result=Get_Post(soapObject);
        if(error.equals(result))
            return null;
        return Json_To_FindComment_ID(result);
    }

    public static FindActivityAllData Get_All_User_Coment_Info_By_ID(int userID,int commentID){
        methodName="Get_All_User_Coment_Info_By_ID";
        soapAction = namespace + "/" + methodName;
        SoapObject soapObject = new SoapObject(namespace, methodName);
        soapObject.addProperty("user", userID);
        soapObject.addProperty("commentID", commentID);
        String result=Get_Post(soapObject);
        if(error.equals(result))
            return null;
        return Json_To_FindActivityAllInformation(result);
    }
    @Nullable
    public static byte[] Get_User_Comment_Image(int id) {
        methodName = "Get_User_Comment_Image";
        soapAction = namespace + "/" + methodName;
        SoapObject soapObject = new SoapObject(namespace, methodName);
        soapObject.addProperty("id", id);
        String result = Get_Post(soapObject);
        if (null == result || error.equals(result)) {
            return null;
        }
        byte[] data = Base64.decode(result);
        return data;
    }

    public static boolean Get_User_Is_Like(int userID, int commentID) {
        methodName = "Get_User_Is_Like";
        soapAction = namespace + "/" + methodName;
        SoapObject soapObject = new SoapObject(namespace, methodName);
        soapObject.addProperty("userID", userID);
        soapObject.addProperty("commentID", commentID);
        String result = Get_Post(soapObject);
        if (success.equals(result)) {
            return true;
        } else {
            return false;
        }
    }

    public static boolean Set_User_Like(int userID, int commentID) {
        methodName = "Set_User_Like";
        soapAction = namespace + "/" + methodName;
        SoapObject soapObject = new SoapObject(namespace, methodName);
        soapObject.addProperty("userID", userID);
        soapObject.addProperty("commentID", commentID);
        String result = Get_Post(soapObject);
        if (success.equals(result)) {
            return true;
        } else {
            return false;
        }
    }

    public static boolean Cancel_User_Like(int userID, int commentID) {
        methodName = "Cancel_User_Like";
        soapAction = namespace + "/" + methodName;
        SoapObject soapObject = new SoapObject(namespace, methodName);
        soapObject.addProperty("userID", userID);
        soapObject.addProperty("commentID", commentID);
        String result = Get_Post(soapObject);
        if (success.equals(result)) {
            return true;
        } else {
            return false;
        }
    }

    public static List<commentReplyData> Get_User_Comment_Reply(int commentID) {
        methodName = "Get_User_Comment_Reply";
        soapAction = namespace + "/" + methodName;
        SoapObject soapObject = new SoapObject(namespace, methodName);
        soapObject.addProperty("commentID", commentID);
        String result = Get_Post(soapObject);
        return Json_To_UserCommentReplyData(result);
    }

    public static int Add_User_Comment_Reply(int userID, int commentID, String replyContent,String intentString) {
        try {
            Thread.sleep(200);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        methodName = "Add_User_Comment_Reply";
        soapAction = namespace + "/" + methodName;
        SoapObject soapObject = new SoapObject(namespace, methodName);
        soapObject.addProperty("userID", userID);
        soapObject.addProperty("commentID", commentID);
        soapObject.addProperty("replyContent", replyContent);
        soapObject.addProperty("intentString",intentString);
        String result = Get_Post(soapObject);
        if (error.equals(result)) {
            return 0;
        } else {
            return Integer.parseInt(result);
        }
    }

    public static boolean Update_User_Comment_Reply(int replyID, String replyContent) {
        methodName = "Update_User_Comment_Reply";
        soapAction = namespace + "/" + methodName;
        SoapObject soapObject = new SoapObject(namespace, methodName);
        soapObject.addProperty("replyID", replyID);
        soapObject.addProperty("replyContent", replyContent);
        String result = Get_Post(soapObject);
        if (success.equals(result)) {
            return true;
        } else {
            return false;
        }
    }

    public static boolean Delete_User_Comment_Reply(int replyID) {
        methodName = "Update_User_Comment_Reply";
        soapAction = namespace + "/" + methodName;
        SoapObject soapObject = new SoapObject(namespace, methodName);
        soapObject.addProperty("replyID", replyID);
        String result = Get_Post(soapObject);
        if (success.equals(result)) {
            return true;
        } else {
            return false;
        }
    }

    public static int Get_User_Comment_Count(int commentID) {
        methodName = "Get_User_Comment_Count";
        soapAction = namespace + "/" + methodName;
        SoapObject soapObject = new SoapObject(namespace, methodName);
        soapObject.addProperty("commentID", commentID);
        String result = Get_Post(soapObject);
        if (error.equals(result)) {
            return 0;
        } else {
            return Integer.parseInt(result);
        }
    }
}

File: app/src/main/java/com/example/sunkai/heritage/connectWebService/HandleFind.kt
package com.example.sunkai.heritage.connectWebService

import android.text.TextUtils
import android.util.Base64
import com.example.sunkai.heritage.activity.loginActivity.LoginActivity
import com.example.sunkai.heritage.entity.CommentReplyInformation
import com.example.sunkai.heritage.entity.UserCommentData
import com.example.sunkai.heritage.entity.UserCommentImages
import com.example.sunkai.heritage.tools.BaiduLocation
import com.example.sunkai.heritage.value.COMMENT_REPLY
import com.example.sunkai.heritage.value.MINI_REPLY
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import okhttp3.FormBody
import java.text.SimpleDateFormat
import java.util.*

/*
 * Created by sunkai on 2018/2/1.
 */
object HandleFind : BaseSetting() {

    fun Add_User_Comment_Information(user_id: Int, comment_title: String, comment_content: String, comment_image: String): Boolean {
        val postUrl = "$URL/AddUserCommentInformation"
        val locaiton = BaiduLocation.location
        val form = FormBody.Builder()
                .add("userID", user_id.toString())
                .add("commentTitle", comment_title)
                .add("commentContent", comment_content)
                .add("commentImage", comment_image)
                .add("location", locaiton)
                .build()
        val result = PutPost(postUrl, form)
        return SUCCESS == result
    }

    private fun handleCommentData(result: String): List<UserCommentData> {
        if (result == ERROR) {
            return arrayListOf()
        } else {
            try {
                val datas = fromJsonToList(result, Array<UserCommentData>::class.java)
                for (data in datas) {
                    data.miniReplys = GetUserCommentReply(data.id, MINI_REPLY)
                }
                return datas
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        return arrayListOf()
    }

    fun GetUserCommentInformation(userID: Int, start: Int = 0): List<UserCommentData> {
        val getUrl = "$URL/GetUserCommentInformation?userID=$userID&start=$start"
        val result = PutGet(getUrl)
        return handleCommentData(result)
    }

    fun GetUserCommentInformationByUser(userID: Int, start: Int = 0): List<UserCommentData> {
        val getUrl = "$URL/GetUserCommentInformationByUser?userID=$userID&start=$start"
        val result = PutGet(getUrl)
        return handleCommentData(result)
    }

    fun GetUserCommentInformaitonByOwn(userID: Int, start: Int = 0): List<UserCommentData> {
        val getUrl = "$URL/GetUserCommentInformaitonByOwn?userID=$userID&start=$start"
        val result = PutGet(getUrl)
        return if (result == ERROR) {
            arrayListOf()
        } else {
            try {
                fromJsonToList(result, Array<UserCommentData>::class.java)
            } catch (e: Exception) {
                e.printStackTrace()
                arrayListOf<UserCommentData>()
            }

        }
    }

    fun GetUserCommentInformationBySameLocation(userID: Int, location: String, start: Int = 0): List<UserCommentData> {
        val getUrl = "$URL/GetUserCommentInformationBySameLocation?userID=$userID&start=$start&location=$location"
        val result = PutGet(getUrl)
        return handleCommentData(result)
    }

    fun GetUserCommentIdByUser(userID: Int): List<UserCommentImages> {
        val getUrl = "$URL/GetUserCommentIdByUser?userID=$userID"
        val result = PutGet(getUrl)
        return if (ERROR == result) {
            arrayListOf()
        } else {
            val idArray = gsonInstance.fromJson(result, IntArray::class.java)
            val resultList = arrayListOf<UserCommentImages>()
            idArray.forEach {
                resultList.add(UserCommentImages(it, HandleFind.GetUserCommentImageUrl(it)))
            }
            resultList
        }
    }

    fun DeleteUserCommentByID(commentID: Int): Boolean {
        val getUrl = "$URL/DeleteUserCommentByID?id=$commentID"
        val result = PutGet(getUrl)
        return result == SUCCESS
    }

    fun UpdateUserCommentImage(commentID: Int, image: ByteArray): Boolean {
        val imgString = Base64.encodeToString(image, Base64.DEFAULT)
        val formBody = FormBody.Builder().add("id", commentID.toString()).add("image", imgString).build()
        val postURL = "$URL/UpdateUserCommentImage"
        val result = PutPost(postURL, formBody)
        return result == SUCCESS
    }

    fun UpdateUserCommentInformaiton(id: Int, title: String, content: String, image: String): String {
        val formBody = FormBody.Builder()
                .add("id", id.toString())
                .add("title", title)
                .add("content", content)
                .add("image", image).build()
        val postUrl = "$URL/UpdateUserCommentInformaiton"
        return PutPost(postUrl, formBody)
    }


    fun GetUserCommentImageUrl(commentID: Int): String {
        val getUrl = "$URL/GetUserCommentImageUrl?id=$commentID"
        val url = PutGet(getUrl)
        if (TextUtils.isEmpty(url) || url == ERROR) {
            return ERROR
        }
        return URL + url
    }

    fun GetAllUserCommentInfoByID(userID: Int, commentID: Int): UserCommentData? {
        val getUrl = "$URL/GetAllUserCommentInfoByID?user=$userID&commentID=$commentID"
        val result = PutGet(getUrl)
        return if (result == ERROR) {
            null
        } else {
            try {
                gsonInstance.fromJson(result, UserCommentData::class.java)
            } catch (e: Exception) {
                e.printStackTrace()
                null
            }
        }
    }

    fun GetCommentLikeNumber(commentID: Int): String {
        val getUrl = "$URL/GetCommentLikeNumber?commentID=$commentID"
        val result = PutGet(getUrl)
        return if (result == ERROR) {
            "0"
        } else {
            result
        }
    }

    fun GetUserCommentCount(commentID: Int, miniReply: Int = COMMENT_REPLY): String {
        val getUrl = "$URL/GetUserCommentCount?commentID=$commentID&miniReply=$miniReply"
        val result = PutGet(getUrl)
        return if (result == ERROR) {
            "0"
        } else {
            result
        }
    }

    fun GetUserCommentReply(commentID: Int, miniReply: Int = COMMENT_REPLY): List<CommentReplyInformation> {
        val getUrl = "$URL/GetUserCommentReply?commentID=$commentID&miniReply=$miniReply"
        val result = PutGet(getUrl)
        return if (result == ERROR) {
            arrayListOf()
        } else {
            try {
                fromJsonToList(result, Array<CommentReplyInformation>::class.java)
            } catch (e: Exception) {
                e.printStackTrace()
                arrayListOf<CommentReplyInformation>()
            }

        }
    }

    fun SetUserLike(userID: Int, commentID: Int): Boolean {
        val getUrl = "$URL/SetUserLike?userID=$userID&commentID=$commentID"
        return PutGet(getUrl) == SUCCESS
    }

    fun CancelUserLike(userID: Int, commentID: Int): Boolean {
        val getUrl = "$URL/CancelUserLike?userID=$userID&commentID=$commentID"
        return PutGet(getUrl) == SUCCESS
    }

    fun AddUserCommentReply(userID: Int, commentID: Int, reply: String, writterID: Int, writterName: String, originalReplyContent: String): String {
        val postUrl = "$URL/AddUserCommentReply"
        val formBody = FormBody.Builder()
                .add("userID", userID.toString())
                .add("commentID", commentID.toString())
                .add("reply", reply)
                .build()
        val result = PutPost(postUrl, formBody)
        if (result != ERROR) {
            if (LoginActivity.userID != 0) {
                GlobalScope.launch {
                    val userName = LoginActivity.userName ?: return@launch
                    val replyTime = Calendar.getInstance().time
                    val timeformat = SimpleDateFormat.getDateInstance().format(replyTime)
                    HandlePush.AddPushMessage(LoginActivity.userID, commentID, writterID, userName, reply, writterName, "20180202", originalReplyContent)
                }
            }
        }
        return result
    }

    fun UpdateUserCommentReply(replyID: Int, reply: String): Boolean {
        val postUrl = "$URL/UpdateUserCommentReply"
        val formBody = FormBody.Builder()
                .add("replyID", replyID.toString())
                .add("reply", reply).build()
        return PutPost(postUrl, formBody) == SUCCESS
    }

    fun DeleteUserCommentReply(replyID: Int): Boolean {
        val getUrl = "$URL/DeleteUserCommentReply?replyID=$replyID"
        return PutGet(getUrl) == SUCCESS
    }

    fun GetUserLikeComment(userID: Int): List<UserCommentData> {
        val getUrl = "$URL/GetUserLikeComment?userID=$userID"
        val result = PutGet(getUrl)
        if (result == ERROR) {
            return arrayListOf()
        }
        try {
            return fromJsonToList(result, Array<UserCommentData>::class.java)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return arrayListOf()
    }

    fun SearchUserCommentInfo(searchInfo: String, userID: Int = LoginActivity.userID): List<UserCommentData> {
        val url = "$URL/SearchUserCommentInfo?userID=$userID&searchInfo=$searchInfo"
        val result = PutGet(url)
        if (result == ERROR) {
            return arrayListOf()
        }
        try {
            return handleCommentData(result)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return arrayListOf()
    }
}
